Python/optimizer_symbols.c

```diff
@@ -48,187 +43,250 @@ static inline int get_lltrace(void) {
-    return true;
+    Py_UNREACHABLE();
+}
+
+static void make_const(JitOptSymbol *sym, PyObject *val)
+{
+    sym->tag = JIT_SYM_KNOWN_VALUE_TAG;
+    sym->value.value = Py_NewRef(val);
 }
 
 void
-_Py_uop_sym_set_const(_Py_UOpsContext *ctx, _Py_UopsSymbol *sym, PyObject *const_val)
+_Py_uop_sym_set_const(JitOptContext *ctx, JitOptSymbol *sym, PyObject *const_val)
 {
-    assert(const_val != NULL);
-    if (sym->flags & IS_NULL) {
-        sym_set_bottom(ctx, sym);
-    }
-    PyTypeObject *typ = Py_TYPE(const_val);
-    if (sym->typ != NULL && sym->typ != typ) {
-        sym_set_bottom(ctx, sym);
-    }
-    if (sym->const_val != NULL) {
-        if (sym->const_val != const_val) {
-            // TODO: What if they're equal?
+    JitSymType tag = sym->tag;
+    switch(tag) {
+        case JIT_SYM_NULL_TAG:
             sym_set_bottom(ctx, sym);
-        }
-    }
-    else {
-        sym_set_flag(sym, NOT_NULL);
-        sym->typ = typ;
-        sym->const_val = Py_NewRef(const_val);
+            return;
+        case JIT_SYM_KNOWN_CLASS_TAG:
+            if (sym->cls.type != Py_TYPE(const_val)) {
+                sym_set_bottom(ctx, sym);
+                return;
+            }
+            make_const(sym, const_val);
+            return;
+        case JIT_SYM_KNOWN_VALUE_TAG:
+            if (sym->value.value != const_val) {
+                Py_CLEAR(sym->value.value);
+                sym_set_bottom(ctx, sym);
+            }
+            return;
+        case JIT_SYM_TUPLE_TAG:
+            sym_set_bottom(ctx, sym);
+            return;
+        case JIT_SYM_TYPE_VERSION_TAG:
+            if (sym->version.version != Py_TYPE(const_val)->tp_version_tag) {
+                sym_set_bottom(ctx, sym);
+                return;
+            }
+            make_const(sym, const_val);
+            return;
+        case JIT_SYM_BOTTOM_TAG:
+            return;
+        case JIT_SYM_NON_NULL_TAG:
+        case JIT_SYM_UNKNOWN_TAG:
+            make_const(sym, const_val);
+            return;
     }
 }
 
 void
-_Py_uop_sym_set_null(_Py_UOpsContext *ctx, _Py_UopsSymbol *sym)
+_Py_uop_sym_set_null(JitOptContext *ctx, JitOptSymbol *sym)
 {
-    if (_Py_uop_sym_is_not_null(sym)) {
+    if (sym->tag == JIT_SYM_UNKNOWN_TAG) {
+        sym->tag = JIT_SYM_NULL_TAG;
+    }
+    else if (sym->tag > JIT_SYM_NULL_TAG) {
         sym_set_bottom(ctx, sym);
     }
-    sym_set_flag(sym, IS_NULL);
 }
 
 void
-_Py_uop_sym_set_non_null(_Py_UOpsContext *ctx, _Py_UopsSymbol *sym)
+_Py_uop_sym_set_non_null(JitOptContext *ctx, JitOptSymbol *sym)
 {
-    if (_Py_uop_sym_is_null(sym)) {
+    if (sym->tag == JIT_SYM_UNKNOWN_TAG) {
+        sym->tag = JIT_SYM_NON_NULL_TAG;
+    }
+    else if (sym->tag == JIT_SYM_NULL_TAG) {
         sym_set_bottom(ctx, sym);
     }
-    sym_set_flag(sym, NOT_NULL);
 }
 
 
-_Py_UopsSymbol *
-_Py_uop_sym_new_unknown(_Py_UOpsContext *ctx)
+JitOptSymbol *
+_Py_uop_sym_new_unknown(JitOptContext *ctx)
 {
     return sym_new(ctx);
```

Comment:
Should this be checked for `NULL` and replaced with `out_of_space` like the others?